### PR TITLE
Changes base image to Alpine Linux with Zulu JRE 8

### PR DESCRIPTION
**Pull request by adriancole**
This change reduces the size of `zipkin-base` from 466.2 MB to 94.2 MB.
This is achieved by switching from `debian:sid` to a minimal base image
built on Alpine Linux; a Java Docker image doesn't get much smaller than
this. Here's what changes.

* `/bin/bash` is `/bin/sh`
* OpenJDK 8 headless to Zulu JRE 8
* Manual install Cassandra as there's no Alpine Linux package
* Works around libraries that load `java.beans` packages not in the JRE
  * Hard-configures Cassandra instead of using snakeyaml
  * Swaps logback for slf4j-simple for Cassandra console logging
* Cleans up python, javadocs, cqlsh as they aren't needed post-install
  * Ad-hoc use of cqlsh can happen off-host

Pulled up `docker-compose up` with these images built locally, ran
`zipkin-tracegen` against them, and saw traces on the web interface.

```bash
# from docker-zipkin
$ docker build -t quay.io/openzipkin/zipkin-base:base-1.5.1 base
$ for I in cassandra collector query web
$   do docker build -t quay.io/openzipkin/zipkin-$I:1.5.1 $I
$ done
$ docker-compose up
# from zipkin
$ bin/tracegen 192.168.99.100
$ open http://192.168.99.100:8080/
```

Sizes before

```bash
$ docker images
REPOSITORY                            TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
quay.io/openzipkin/zipkin-web         1.5.1               febb7c1c8e09        7 hours ago         515.6 MB
quay.io/openzipkin/zipkin-cassandra   1.5.1               ca376e0a1eb6        7 hours ago         618.2 MB
quay.io/openzipkin/zipkin-query       1.5.1               e4b5209e85dd        7 hours ago         525 MB
quay.io/openzipkin/zipkin-collector   1.5.1               ac8bbc05c2e4        7 hours ago         533.6 MB
```

```bash
$ docker images
REPOSITORY                            TAG                 IMAGE ID            CREATED              VIRTUAL SIZE
quay.io/openzipkin/zipkin-web         1.5.1               05b4d0d6b886        49 seconds ago       143.5 MB
quay.io/openzipkin/zipkin-query       1.5.1               cedcdb8d1fba        About a minute ago   153 MB
quay.io/openzipkin/zipkin-collector   1.5.1               a55c2f3febde        About a minute ago   161.6 MB
quay.io/openzipkin/zipkin-cassandra   1.5.1               2fc3b7993288        About a minute ago   179.9 MB
quay.io/openzipkin/zipkin-base        base-1.5.1          c1174dc1512b        2 minutes ago        94.17 MB
delitescere/java                      8                   8effc79cfdc8        19 minutes ago       90.43 MB
```

### CODE REVIEW

**Comment by adriancole** [comment]
This work was almost all @abesto. The last mile was far easier with advice from @michaelsembwever

File: cassandra/README.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+This changes the datastax distribution so that it doesn't require access to the `javax.beans` package.
+
+```bash
+# compiling a configuration loader which doesn't use snakeyaml
+$ javac -classpath dsc-cassandra-2.1.9/lib/apache-cassandra-2.1.9.jar ZipkinConfigurationLoader.java
```
Comment (abesto) [inline cassandra/README.md]: How about something like

```bash
CASSANDRA_VERSION=2.1.9
curl -L http://downloads.datastax.com/community/dsc-cassandra-$CASSANDRA_VERSION-bin.tar.gz | tar xz
javac -classpath dsc-cassandra-2.1.9/lib/apache-cassandra-2.1.9.jar ZipkinConfigurationLoader.java
git add  ZipkinConfigurationLoader.class
rm -rf dsc-cassandra-2.1.9
```

File: cassandra/ZipkinConfigurationLoader.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      parameters.put("seeds", ip);
+      config.seed_provider.parameters = parameters;
+      return config;
+    } catch (SocketException e) {
+      throw new ConfigurationException("couldn't get host ip", e);
+    }
+  }
+
+  // http://stackoverflow.com/questions/8765578/get-local-ip-address-without-connecting-to-the-internet
+  private static String getHostIpWithoutNSLookup() throws SocketException {
```
Comment (abesto) [inline cassandra/ZipkinConfigurationLoader.java]: Maybe add a comment that this is, and is intended to be, the ip of the docker container (as opposed to the underlying docker host)

File: cassandra/ZipkinConfigurationLoader.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    return Collections.list(NetworkInterface.getNetworkInterfaces()).stream()
+        .flatMap(i -> Collections.list(i.getInetAddresses()).stream())
+        .filter(ip -> ip instanceof Inet4Address && ip.isSiteLocalAddress())
+        .findAny().orElseThrow(SocketException::new)
+        .getHostAddress();
+  }
+
+  private static Config fromConfigYaml() {
```
Comment (abesto) [inline cassandra/ZipkinConfigurationLoader.java]: This name makes me expect that stuff will be loaded from a yml file, which is exactly what it doesn't do :D How about calling it `defaultConfig` or just adding a comment to explain?

**Comment by abesto** [comment]
This is so exciting :)

Couple of nits inline. The only big one: we need to check in the compiled `.class` file to make this work on quay.io

File: cassandra/README.md (hunk 1 above)
Comment (adriancole) [inline cassandra/README.md]: sure

**Comment by adriancole** [comment]
hehe the compiled file is checked in :) agree we need that as the whole point is to have this work with JRE. Addressing notes now.

**Comment by adriancole** [comment]
great point about comments.. done

File: cassandra/ZipkinConfigurationLoader.java (hunk 2 above)
Comment (adriancole) [inline cassandra/ZipkinConfigurationLoader.java]: done

File: cassandra/ZipkinConfigurationLoader.java (hunk 3 above)
Comment (adriancole) [inline cassandra/ZipkinConfigurationLoader.java]: done

**Comment by abesto** [comment]
Ah, sorry, guess I missed the tiny line for the binary file in the diff view.

**Comment by abesto** [comment]
:sparkles: Looks good! :shipit:

**Comment by adriancole** [comment]
ps to those wanting to get rid of the UnsatisfiedLinkError in the log. here's the cassandra upgrade https://github.com/openzipkin/docker-zipkin/pull/48

**Comment by adriancole** [comment]
Also.. This allows docker-compose to work on EC2's default of a t1.micro! IOTW zipkin for 1c/hr
cc @spencergibb

**Comment by spencergibb** [comment]
:+1:

**Comment by spencergibb** [comment]
Interestingly, I tried the java docker image with a spring-cloud/boot app and had to use `delitescere/jdk:8` for some reason. /cc @dsyer @mstine

**Pull request merged**